### ISSUE TITLE
WebGL 2 not available

### ISSUE THREAD

**Issue by starry-abyss** [godotengine/godot#12416]
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
Windows 7 64bit, Godot 1063ddf, GTX 660, NVIDIA driver 388.00, Firefox 56.0 (64 bit)

**Issue description:**
I expected it to work across browsers and systems, since this is what HTML export is about.
But I see the error message instead:

![default](https://user-images.githubusercontent.com/11571820/32054660-db5131c8-ba67-11e7-9a8f-8f15ffe3ffb9.png)

**Steps to reproduce:**
1. Export an empty project to javascript or webassembly
2. Open .html file in Firefox

**Comment by akien-mga** [comment]
What does http://webglreport.com/?v=2 say?

@eska014 Could we get more debug output as to why WebGL 2 is not available, or is it the same kind of try and error situation as for OpenGL on desktop?

**Comment by starry-abyss** [comment]
![default](https://user-images.githubusercontent.com/11571820/32056433-59947158-ba6d-11e7-9321-c857b8ccb7a7.png)

**Comment by eska014** [comment]
@starry-abyss Are your GPU drivers up to date? Also make sure `webgl.enable-webgl2` is enabled in `about:config`

@akien-mga I'm afraid a "`try {}` and `Error()` situation" is exactly what it is: https://github.com/godotengine/godot/blob/2987e6ce4ffc69da5efc0903cc00ce12e72d2125/platform/javascript/engine.js#L129-L134
We could check browser and browser version to report if it should be supported, that's probably what webglreport.com is doing as well.

**Comment by starry-abyss** [comment]
Yes, I just installed newest drivers from NVIDIA's site. They reject to install from Windows Update somehow, but NVIDIA Control Panel says: 388.00.

**Comment by akien-mga** [comment]
>  We could check browser and browser version to report if it should be supported, that's probably what webglreport.com is doing as well.

AFAIK webglreport.com really checks the capabilities of your current setup, I've had it report lacking support in Firefox Nightly due to sandboxing [a few months ago](https://bugzilla.mozilla.org/show_bug.cgi?id=1385715), and the information it shows for me are linked to my system, not only the browser (D3D11, Intel HD 5500, etc.):
![image](https://user-images.githubusercontent.com/4701338/32057404-68e30968-ba67-11e7-9b81-d8cf7647f67e.png)

**Comment by akien-mga** [comment]
Source for webglreport.com: https://github.com/AnalyticalGraphicsInc/webglreport/blob/master/webglreport.js

**Comment by starry-abyss** [comment]
Chrome 62.0.3202.62 shows the same for me at webglreport.com
Can't open the build either, the loading wheel just infinitely loops (probably Chrome needs some local server because of security, which I don't know how to get running quickly?)

**Comment by eska014** [comment]
What does the [WebGL 1 tab](http://webglreport.com/?v=1) report?
Is `webgl.disabled` off in `about:config`? What results do you get if you enable `webgl.force-enabled`?

>probably Chrome needs some local server because of security, which I don't know how to get running quickly?

https://godot.readthedocs.io/en/latest/learning/workflow/export/exporting_for_web.html#starting-exported-games-from-the-local-file-system

**Comment by starry-abyss** [comment]
WebGL 1 is reported as supported on both browsers.
Assuming other questions are Firefox-specific: webgl.disabled = false, and with webgl.force-enabled the webgl 2 report is still red

**Comment by starry-abyss** [comment]
Is the ANGLE thing something installed or updated separately? Does the ANGLE thing mean I'm running on Intel instead of NVIDIA?

![default](https://user-images.githubusercontent.com/11571820/32062257-a691846e-ba7c-11e7-82ba-ad55722a9daa.png)

**Comment by Calinou** [comment]
To get a Web server running, the `serve` npm package works pretty well.

```
# Install the package, you need to install Node.js which includes npm
npm install -g serve

# Run this command in the export directory
serve
```

**Comment by eska014** [comment]
No, browsers implement WebGL on Windows per Angle, but this is included with the browsers

**Comment by starry-abyss** [comment]
This is what Chrome says about my GPU, if helps:

[gpu.zip](https://github.com/godotengine/godot/files/1424042/gpu.zip)

**Comment by starry-abyss** [comment]
In Chrome:

![default](https://user-images.githubusercontent.com/11571820/32317238-324fde00-bfc4-11e7-913d-6855b727f401.png)

So, I have no idea how to run the exported project on my machine (don't even say about different players' configurations), with browsers and drivers updated.

**Comment by mhilbrunner** [comment]
This doesn't seem to be something that can be fixed on Godot's end? If other tools don't detect WebGL 2 support, there is probably not much we can do, close this?

**Comment by starry-abyss** [comment]
I guess that if Godot enforces WebGL version 2, then there should be some manual on how to get it running. Simply updating everything doesn't work. If it doesn't for me, then how can I make it work on hundreds of players' computers?

**Comment by mhilbrunner** [comment]
As far as I can see, WebGL version 2 support is available in recent versions of most major browsers.
If this doesn't work for you, this is something that you should probably take up with your browser, your OS, or graphcis hardware/driver vendor.

If you release a game on this platform, you should have some explanations for your users available either surrounding the embedded game or on your home page / some supporting page.

But WebGL2 not working on your system isn't something Godot can do anything about.

Releases supporting WebGL2:
    Firefox release 51
    Chrome for desktop release 56
    Chrome for Android release 58
    Android WebView release 58
    Opera release 43
    Opera for Android release 43

**Comment by starry-abyss** [comment]
> WebGL version 2 support is available in recent versions of most major browsers

It's available on paper, but practically it doesn't work.

> take up with your browser, your OS, or graphcis hardware/driver vendor.

Everything of this was updated. Didn't help.

> you should have some explanations for your users available either surrounding the embedded game or on your home page / some supporting page

I can't provide such information, because I even can't get it running myself.

> But WebGL2 not working on your system isn't something Godot can do anything about.

If it's working on several computers of Godot maintainters, doesn't mean that Godot has a web export available for spread use.

> Releases supporting WebGL2:
> Firefox release 51
> Chrome for desktop release 56

WebGL 2 doesn't work for me neither in Chrome 63 nor in Firefox 57

**Comment by mhilbrunner** [comment]
If WebGL 2 doesn't work for you at all, the only thing Godot *could* do would be to support WebGL 1. I don't know if that is something that would be done; maybe rename your issue to "Add support for WebGL 1".

Your issue, as is, states: "WebGL 2 not available". That isn't a Godot issue, plain and simple. Its one of your setup.

**Comment by starry-abyss** [comment]
> I don't know if that is something that would be done

Neither do I, but the problem as I see it is serious. I'm not here for joking jokes or getting someone angry. Maybe someone else knows what can be done.

> maybe rename your issue to "Add support for WebGL 1".

I can't just decide that this is the only possible solution.

> Your issue, as is, states: "WebGL 2 not available"

It's the closest searchable short description I can think of.

> That isn't a Godot issue, plain and simple

If web export isn't working, it's Godot issue. But since I'm not a web developer at all, I can't do deep investigations on my own (tips from googling forums didn't help) and propose solutions. Saying that I'm not a web developer means exactly that I don't even plan developing web games with Godot except for jam ones.

To conclude, I'm leaving the issue open and as is, until someone competent in the field decides what to do with it.

**Comment by mhilbrunner** [comment]
I'm neither joking nor angry either :)

>If web export isn't working, it's Godot issue.

The web export isn't working on your setup because WebGL 2 is not working on your setup. There isn't anything to investigate. There are only two possible options I see to get it to work for you:

1. You get WebGL 2 to work
2. Godot adds support for WebGL 1

**Comment by akien-mga** [comment]
Well don't fight guys. 

First of all, no, WebGL 1 won't be added.

Then, @starry-abyss has the specs to support WebGL 2 in theory: recent Nvidia drivers, decent GPU, latest Firefox and Chrome.

And still it's not working, so there's an issue on your system, and this has to be discussed with either your drivers vendors or your browsers vendors. It's fully unrelated to Godot itself - we're just users of the WebGL 2 technology provided by browsers.

To make an analogy, if your keyboard drivers were broken and you were not able to pass any input to any program, it wouldn't be a Godot bug either, even if it prevents you from using Godot. In your case, you can't use any web application using WebGL 2 - Godot happens to be the first one problematic for you, but that doesn't make it a Godot issue.

I'd suggest you open a bug report against Firefox on http://bugzilla.mozilla.org/ - the Mozilla folk are very helpful and will help you debug why WebGL 2 isn't working on your a priori capable system.

**Comment by vnen** [comment]
This can't be solved from Godot's side, it's a fault of the browsers. The WebGL Report shows ANGLE with D3D9, which does not support GLES 3.0 / WebGL 2. It's only supported with D3D11. Might be a driver's fault too, but nothing we can do about it.

**Comment by BrutPitt** [comment]
Try to disable ANGLE on Firefox:
In "about:config"
webgl.disable-angle -> TRUE
So You use native OpenGL.

I have same trouble on Windows 10 PC: Firefox and chromium based browser (Chrome and Vivaldi), using ANGLE, fail to create WebGL2 context.
FireFox, JUST and ONLY disabling ANGLE, works fine (for me).
On Linux, with native OpenGL support, everything is fine... on all browser.

**Comment by starry-abyss** [comment]
@BrutPitt Thank you, that totally worked on Firefox! Now to figure out how to it can be fixed (I'm not going to enforce the workarounds on players)!

**Comment by BrutPitt** [comment]
@starry-abyss 

After a night spent in updates... windows and drivers... without solution... I solved this problem (for me)... in another way.

My problem was with dx11/ANGLE initialization: (eglInitialize() error: egl_not_initialized... displayed in "chrome://gpu" information page)

At end I solved using DDU (Display Driver Uninstaller, freeware utility, last version), to remove any driver registry occurrence and dll, launched with windows in "save mode".
I recommend you, if like me you have installed drivers of other brands (AMD, Intel), to repeat the procedure also for any other drivers brand.

After reboot, I have reinstalled the drivers, and now all works fine.
